server/backends/migration_cache/migration_cache.go

```diff
@@ -232,3 +250,54 @@ func (mc *MigrationCache) Set(ctx context.Context, d *repb.Digest, data []byte)
 
 	return srcErr
 }
+
+func (mc *MigrationCache) copyDataInBackground(ctx context.Context) error {
+	for {
+		select {
+		case <-mc.quitChan:
+			return nil
+		case digestToCopy := <-mc.copyChan:
+			alreadyCopied, err := mc.dest.Contains(ctx, digestToCopy)
+			if err != nil {
+				log.Warningf("Migration copy err: Could not call Contains on dest cache: %s", err)
+				continue
+			}
+
+			if !alreadyCopied {
+				srcReader, err := mc.src.Reader(ctx, digestToCopy, 0, 0)
+				if err != nil {
+					if !status.IsNotFoundError(err) {
+						log.Warningf("Migration copy err: Could not create reader for digest %v from src cache: %s", digestToCopy, err)
+					}
+					continue
+				}
+
+				destWriter, err := mc.dest.Writer(ctx, digestToCopy)
+				if _, err = io.Copy(destWriter, srcReader); err != nil {
+					log.Warningf("Migration copy err: Could not create writer for digest %v to dest cache: %s", digestToCopy, err)
+				}
+				if err := destWriter.Close(); err != nil {
```

Comment:
In our current Writer implementations, Close really means "Commit write" so we should not call it in in the error case. I think Tyler was going to fix this soon so that we have a separate Commit() and Close() methods.

Reply:
Ah I see - added an early return